### PR TITLE
Cache actions and events queries between event processings

### LINKED ISSUE #809: Improve event insert/worker performance
**Describe the bug**
We have a user that's doing 40 req/sec. The workers can't keep up with the volume of events, so the Redis cache fills up and then falls over

At the moment their setup is
1. 3 web dynos 2X
1. 15 worker dynos 2X
1. 1GB Redis cache
1. Standard-1 postgres database

As you can see [in this test](https://github.com/PostHog/posthog/blob/master/posthog/tasks/test/test_process_event.py#L19), we do 19 queries for each event insert (with autocapture).

**ideas**
- I think the query that [finds the relevant actions for an event](https://github.com/PostHog/posthog/blob/master/posthog/models/event.py#L239) is slow, so it might be worth using the same pattern we have for Cohorts to calculate them after the fact.
- For this specific user, the dom tree is probably unique for a lot of events so have to be re-inserted. We already bulk-insert all of the Elements, but there might still be gains to be had here

### PR DESCRIPTION
**Pull request by fuziontech**
## Changes
Cache sql and params between event processing. Building the query takes some cpu time so this speeds things up to about 70ms per event from 180ms per event before on my laptop.

## Checklist
- [x] All querysets/queries filter by Team (if applicable)
- [x] Backend tests (if applicable)
- [x] Cypress E2E tests (if applicable)

### CODE REVIEW

File: posthog/models/event.py [hunk 1]
```diff
@@ -237,35 +246,58 @@ def person(self):
...
-        for action in actions:
-            events = events.annotate(
-                **{
-                    "action_{}".format(action.pk): Event.objects.filter(pk=self.pk)
-                    .query_db_by_action(action)
-                    .values("id")[:1]
-                }
-            )
-        event = [event for event in events][0]
+        if not self.team_id in LAST_UPDATED_TEAM_ACTION \
```
Comment (timgl) [inline posthog/models/event.py:262]: Maybe personal preference but I think these two if statements could be split out to different functions

File: posthog/models/event.py [hunk 2]
```diff
@@ -237,35 +246,58 @@ def person(self):
...
+                events = events.annotate(
+                    **{
+                        "action_{}".format(action.pk): Event.objects.filter(pk=self.pk)
+                        .query_db_by_action(action)
+                        .values("id")[:1]
+                    }
+                )
+            q, p = events.query.sql_with_params()
+            qp = tuple(['%s' if i == self.pk else i for i in p])
+            qcache = {self.event: (q, qp)}
```
Comment (timgl) [inline posthog/models/event.py:282]: These 2 lines are a bit magic (especially with the non-descriptive variable names). Is there a way of doing this that's slightly less magic or just more descriptive?

File: posthog/models/event.py [hunk 3]
```diff
@@ -237,35 +246,58 @@ def person(self):
...
-                    .query_db_by_action(action)
-                    .values("id")[:1]
-                }
-            )
-        event = [event for event in events][0]
+        if not self.team_id in LAST_UPDATED_TEAM_ACTION \
+            or not self.team_id in TEAM_EVENT_ACTION_QUERY_CACHE \
+            or not self.event in TEAM_EVENT_ACTION_QUERY_CACHE[self.team_id] \
+            or not self.team_id in TEAM_ACTION_QUERY_CACHE \
+            or (last_updated_action_ts > LAST_UPDATED_TEAM_ACTION[self.team_id]):
```
Comment (timgl) [inline posthog/models/event.py:266]: Is this logic implicitly tested by the current tests, or should we add another test to test this logic specifically? I think we do have some tests that test updating of actions, just want to make sure they would hit this. Maybe a specific test would be good?

Comment (fuziontech) [reply posthog/models/event.py:266]: I think a specific test would be the right way to do this. That's what I was going to hack on a little later if I have time.

File: posthog/models/event.py (hunk 2 above)
Comment (fuziontech) [reply posthog/models/event.py:282]: Totally agree. I'll work that in too.

File: posthog/models/event.py (hunk 1 above)
Comment (fuziontech) [reply posthog/models/event.py:262]: I think I know what you are saying here. I'll update PR and see if we are on same page

**Comment by timgl** [comment]
@fuziontech Sorry I missed this. Looks like the rebase went a bit awry. once that's done I'll qa and merge :)

**Comment by fuziontech** [comment]
Woops!

**Comment by fuziontech** [comment]
Cleaned this up a bit. I had a really hard time figuring out why tests broke after rebasing. After quite a bit of debugging I changed the `BaseTest` from `TestCase` to [TransactionTestCase](https://docs.djangoproject.com/en/3.0/topics/testing/tools/#transactiontestcase) which is the parent class for BaseTest and forces the DB to commit changes before moving on instead of rolling back each Test in a transaction. The downside for this is that after each test under this parent class all tables are truncated. This slows down testing significantly. I happened upon this because I was running out of leads and enabled the `TransactionTestCase` just so that I could inspect the data in the test DB.

TODO for me is to research and understand why these tests fail under `TestCase` and succeed under `TransactionTestCase`.

**Pull request merged**